### PR TITLE
controller-manager: switch to options+config pattern and add https+auth

### PR DESCRIPTION
**Pull request by sttts**
This PR switch the {kube,cloud}-controller-managers to use the Options+Config struct pattern for bootstrapping, as we use it throughout all apiservers. This allows us to easily plug in https and authn/z support.

Fixes parts of https://github.com/kubernetes/kubernetes/issues/59483

This is equivalent to https://github.com/kubernetes/kubernetes/pull/59408 after squashing.

```release-note
Deprecate insecure HTTP port of kube-controller-manager and cloud-controller-manager. Use `--secure-port` and `--bind-address` instead.
```

### CODE REVIEW

**Comment by stewart-yu** [comment]
Great! Test case failed seems related to `--port`  and `--adress` flag, i will fix it in my local first. @sttts

**Comment by sttts** [comment]
First commit shouldn't be there. Will rebase again...

File: cmd/controller-manager/app/options/options.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	genericcontrollermanager "k8s.io/kubernetes/cmd/controller-manager/app"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+	"k8s.io/kubernetes/pkg/client/leaderelectionconfig"
+)
+
+// GenericControllerManagerOptions is the common structure for a controller manager. It works with NewGenericControllerManagerOptions
+// and AddDefaultControllerFlags to create the common components of kube-controller-manager and cloud-controller-manager.
+type GenericControllerManagerOptions struct {
+	ComponentConfig componentconfig.KubeControllerManagerConfiguration
```
Comment (deads2k) [inline cmd/controller-manager/app/options/options.go]: Add a TODO, this isn't generic and needs removal

File: cmd/controller-manager/app/options/options.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+		ReconcilerSyncLoopPeriod:              metav1.Duration{Duration: 60 * time.Second},
+		EnableTaintManager:                    true,
+		HorizontalPodAutoscalerUseRESTClients: true,
+	}
+}
+
+// AddFlags adds common/default flags for both the kube and cloud Controller Manager Server to the
+// specified FlagSet. Any common changes should be made here. Any individual changes should be made in that controller.
+func (o *GenericControllerManagerOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&o.ComponentConfig.UseServiceAccountCredentials, "use-service-account-credentials", o.ComponentConfig.UseServiceAccountCredentials, "If true, use individual service account credentials for each controller.")
```
Comment (deads2k) [inline cmd/controller-manager/app/options/options.go]: separate all this. It isn't generic and needs to be removed.

File: cmd/controller-manager/app/serve.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/util/configz"
+)
+
+type serveFunc func(handler http.Handler, shutdownTimeout time.Duration, stopCh <-chan struct{}) error
+
+// Serve creates a base handler chain for a controller manager. It runs the
+// the chain with the given serveFunc.
+func Serve(c *CompletedConfig, serveFunc serveFunc, stopCh <-chan struct{}) error {
```
Comment (deads2k) [inline cmd/controller-manager/app/serve.go]: Can we factor this as a handlerchain building func for the apiserver?

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -174,13 +161,13 @@ func Run(s *options.CloudControllerManagerServer) error {
...
-	rl, err := resourcelock.New(s.LeaderElection.ResourceLock,
+	rl, err := resourcelock.New(c.Generic.ComponentConfig.LeaderElection.ResourceLock,
 		"kube-system",
 		"cloud-controller-manager",
-		leaderElectionClient.CoreV1(),
+		c.Generic.Extra.LeaderElectionClient.CoreV1(),
 		resourcelock.ResourceLockConfig{
 			Identity:      id,
-			EventRecorder: recorder,
+			EventRecorder: c.Generic.Extra.EventRecorder,
```
Comment (deads2k) [inline cmd/cloud-controller-manager/app/controllermanager.go]: TODO this indicates that the eventrecorder isn't extra.

File: cmd/controller-manager/app/config.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// Config is the main context object for the controller manager.
+type Config struct {
+	ComponentConfig componentconfig.KubeControllerManagerConfiguration
+	SecureServing   *apiserver.SecureServingInfo
+	InsecureServing *InsecureServingInfo
+	Authentication  apiserver.AuthenticationInfo
+	Authorization   apiserver.AuthorizationInfo
+	Extra           ExtraConfig
```
Comment (deads2k) [inline cmd/controller-manager/app/config.go]: The concept of extra seems a little weird here.  If this is generic, no one will ever add to it.

Comment (deads2k) [reply cmd/controller-manager/app/config.go]: The struct separation is good, the name looks off.

File: cmd/controller-manager/app/options/insecure_serving.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	// MarkDeprecated hides the flag from the help. We don't want that:
+	// fs.MarkDeprecated("address", "see --bind-address instead.")
+
+	fs.IntVar(&s.BindPort, "port", s.BindPort, "The port on which to serve HTTP insecurely without authentication and authorization. If 0, don't serve HTTPS at all. DEPRECATED: see --secure-port instead.")
+	// MarkDeprecated hides the flag from the help. We don't want that:
+	// fs.MarkDeprecated("port", "see --secure-port instead.")
+}
+
+// ApplyTo adds InsecureServingOptions to the insecureserverinfo amd kube-controller manager configuration.
+func (s *InsecureServingOptions) ApplyTo(c **genericcontrollermanager.InsecureServingInfo, cfg *componentconfig.KubeControllerManagerConfiguration) error {
```
Comment (deads2k) [inline cmd/controller-manager/app/options/insecure_serving.go]: double pointer reads like a bug.

File: cmd/controller-manager/app/options/options.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+	return o
+}
+
+// NewDefaultControllerManagerComponentConfig returns default kube-controller manager configuration object.
+func NewDefaultControllerManagerComponentConfig(insecurePort int32) componentconfig.KubeControllerManagerConfiguration {
+	return componentconfig.KubeControllerManagerConfiguration{
+		Controllers:                                     []string{"*"},
+		Port:                                            insecurePort,
+		Address:                                         "0.0.0.0",
```
Comment (deads2k) [inline cmd/controller-manager/app/options/options.go]: I didn't expect to see this here.  Isn't it replaced?

File: cmd/controller-manager/app/serve.go (hunk 3 above)
Comment (deads2k) [reply cmd/controller-manager/app/serve.go]: > Can we factor this as a handlerchain building func for the apiserver?

followup is fine

File: pkg/volume/fc/fc.go [hunk 8]
```diff
@@ -82,11 +82,7 @@ func (plugin *fcPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
 }
 
 func (plugin *fcPlugin) CanSupport(spec *volume.Spec) bool {
-	if (spec.Volume != nil && spec.Volume.FC == nil) || (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FC == nil) {
-		return false
-	}
-
-	return true
+	return (spec.Volume != nil && spec.Volume.FC != nil) || (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.FC != nil)
```
Comment (deads2k) [inline pkg/volume/fc/fc.go]: why the format change? How did you even end up herE?

**Comment by deads2k** [comment]
Looks like a move in the right direction.

@k8s-mirror-api-machinery-pr-reviews 
/assign @cheftako 

Added Walter to make sure he see it.

File: cmd/controller-manager/app/options/options.go (hunk 7 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: I want to remove port and address here. Wasn't sure about the status as an api and did not remove it yet.

File: cmd/controller-manager/app/options/insecure_serving.go (hunk 6 above)
Comment (sttts) [reply cmd/controller-manager/app/options/insecure_serving.go]: actually it isn't. We have to be able to set a new value to a pointer without referencing the outer struct. The leads to a double pointer. Not common in Go. You haven't written C for quite some time :)

File: cmd/controller-manager/app/config.go (hunk 5 above)
Comment (sttts) [reply cmd/controller-manager/app/config.go]: Agree.

File: pkg/volume/fc/fc.go (hunk 8 above)
Comment (stewart-yu) [reply pkg/volume/fc/fc.go]: seems catch other merged commit log[https://github.com/kubernetes/kubernetes/pull/59548](https://github.com/kubernetes/kubernetes/pull/59548)

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 9]
```diff
@@ -20,32 +20,24 @@ import (
 	"fmt"
 	"math/rand"
 	"net"
-	"net/http"
-	"net/http/pprof"
```
Comment (hzxuzhonghu) [inline cmd/cloud-controller-manager/app/controllermanager.go]: where is the pprof now?

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 10]
```diff
@@ -83,84 +81,73 @@ the cloud specific control loops shipped with Kubernetes.`,
...
-		return time.Duration(float64(s.MinResyncPeriod.Nanoseconds()) * factor)
+		return time.Duration(float64(c.Generic.ComponentConfig.MinResyncPeriod.Nanoseconds()) * factor)
 	}
 }
 
 // Run runs the ExternalCMServer.  This should never exit.
-func Run(s *options.CloudControllerManagerServer) error {
-	if s.CloudProvider == "" {
-		glog.Fatalf("--cloud-provider cannot be empty")
-	}
```
Comment (stewart-yu) [inline cmd/cloud-controller-manager/app/controllermanager.go]: should we remove thoes line?

File: cmd/controller-manager/app/options/options.go (hunk 7 above)
Comment (deads2k) [reply cmd/controller-manager/app/options/options.go]: > I want to remove port and address here. Wasn't sure about the status as an api and did not remove it yet.

ok.  Note it in your description as a followup please.

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 10 above)
Comment (deads2k) [reply cmd/cloud-controller-manager/app/controllermanager.go]: > should we remove thoes line?

they just moved to validation where they belonged.

File: cmd/controller-manager/app/options/options.go (hunk 2 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: This is a follow-up.

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 9 above)
Comment (sttts) [reply cmd/cloud-controller-manager/app/controllermanager.go]: It moved into the generic `Serve` func.

File: cmd/controller-manager/app/options/options.go (hunk 7 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: Added follow-up to https://github.com/kubernetes/kubernetes/issues/59483.

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 4 above)
Comment (sttts) [reply cmd/cloud-controller-manager/app/controllermanager.go]: done

File: cmd/controller-manager/app/config.go (hunk 5 above)
Comment (sttts) [reply cmd/controller-manager/app/config.go]: done

File: cmd/controller-manager/app/serve.go (hunk 3 above)
Comment (sttts) [reply cmd/controller-manager/app/serve.go]: Added follow-up to https://github.com/kubernetes/kubernetes/issues/59483.

File: cmd/controller-manager/app/options/options.go (hunk 2 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: Added follow-up to https://github.com/kubernetes/kubernetes/issues/59483.

File: cmd/controller-manager/app/options/options.go (hunk 1 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: done

**Comment by sttts** [comment]
Addressed all comments. Ptal.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (cheftako) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go]: 2016?

File: pkg/master/ports/ports.go [hunk 12]
```diff
@@ -26,12 +26,12 @@ const (
 	// SchedulerPort is the default port for the scheduler status server.
 	// May be overridden by a flag at startup.
 	SchedulerPort = 10251
-	// ControllerManagerPort is the default port for the controller manager status server.
+	// InsecureKubeControllerManagerPort is the default port for the controller manager status server.
 	// May be overridden by a flag at startup.
-	ControllerManagerPort = 10252
```
Comment (cheftako) [inline pkg/master/ports/ports.go]: Given that out of tree providers may be linking against this, its probably worth noting in the release notes.

**Review by cheftako — COMMENTED** [review]
Thank you @sttts for taking care of this. I realize we have a bunch more to do in this area but its a great start.

File: cmd/controller-manager/app/config.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import (
+	apiserver "k8s.io/apiserver/pkg/server"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+)
+
+// ExtraConfig are part of Config, also can place you custom config here.
+type ExtraConfig struct {
```
Comment (deads2k) [inline cmd/controller-manager/app/config.go]: Where is this used now?  I don't object to gathering.  It just read weird.  I don't really care at the moment, but no cruft.

File: cmd/controller-manager/app/options/options.go (hunk 2 above)
Comment (deads2k) [reply cmd/controller-manager/app/options/options.go]: > This is a follow-up.

accepted

File: cmd/controller-manager/app/serve.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+		if c.ComponentConfig.EnableContentionProfiling {
+			goruntime.SetBlockProfileRate(1)
+		}
+	}
+	configz.InstallHandler(mux)
+	mux.Handle("/metrics", prometheus.Handler())
+
+	requestContextMapper := apirequest.NewRequestContextMapper()
+	requestInfoResolver := &apirequest.RequestInfoFactory{}
+	failedHandler := genericapifilters.Unauthorized(requestContextMapper, legacyscheme.Codecs, c.Authentication.SupportsBasicAuth)
```
Comment (deads2k) [inline cmd/controller-manager/app/serve.go]: This should never support basic auth

File: pkg/master/ports/ports.go (hunk 12 above)
Comment (deads2k) [reply pkg/master/ports/ports.go]: > Given that out of tree providers may be linking against this, its probably worth noting in the release notes.

We should kill these constants.  Experience in vendoring now says that no one in their right mind should bring in kube for a constant like this.

**Comment by deads2k** [comment]
Looks like its down to nits from myself and @cheftako .  Please fix them up as you rebase (conflicts) and squash to something good for history.

/lgtm

File: cmd/cloud-controller-manager/app/config/config.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+package app
+
+import (
+	"time"
+
+	genericcontrollermanager "k8s.io/kubernetes/cmd/controller-manager/app"
+)
+
+// ExtraConfig are part of Config, also can place you custom config here.
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/cloud-controller-manager/app/config/config.go]: done

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 16]
```diff
@@ -83,84 +81,73 @@ the cloud specific control loops shipped with Kubernetes.`,
...
 
-		if err := StartControllers(s, kubeconfig, rootClientBuilder, clientBuilder, stop, recorder, cloud); err != nil {
+		if err := startControllers(c, c.Generic.Kubeconfig, rootClientBuilder, clientBuilder, stop, c.Generic.EventRecorder, cloud); err != nil {
 			glog.Fatalf("error running controllers: %v", err)
 		}
 	}
 
-	if !s.LeaderElection.LeaderElect {
+	if !c.Generic.ComponentConfig.LeaderElection.LeaderElect {
 		run(nil)
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/cloud-controller-manager/app/controllermanager.go]: Leaving this to https://github.com/kubernetes/kubernetes/pull/57932.

File: cmd/controller-manager/app/options/insecure_serving.go [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+
+// AddDeprecatedFlags adds deprecated flags related to insecure serving for controller manager to the specified FlagSet.
+// TODO: remove it until kops stop using `--address`
+func (s *InsecureServingOptions) AddDeprecatedFlags(fs *pflag.FlagSet) {
+	if s == nil {
+		return
+	}
+
+	fs.IPVar(&s.BindAddress, "address", s.BindAddress,
+		"The IP address on which to listen for the --port port. DEPRECATED: see --bind-address instead.")
```
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/controller-manager/app/options/insecure_serving.go]: +1

[the remark this answers is not in the document]
Comment (sttts) [reply cmd/controller-manager/app/options/insecure_serving.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go (hunk 11 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go]: fixed

File: pkg/master/ports/ports.go (hunk 12 above)
Comment (sttts) [reply pkg/master/ports/ports.go]: > Given that out of tree providers may be linking against this, its probably worth noting in the release notes.

We usually do not document dev-only changes, which reminds me that we really want `dev-release-note` blocks in the descriptions and collect them for a dev changelog on release.

File: cmd/controller-manager/app/config.go (hunk 13 above)
Comment (sttts) [reply cmd/controller-manager/app/config.go]: left over. Removed.

File: cmd/controller-manager/app/serve.go (hunk 14 above)
Comment (sttts) [reply cmd/controller-manager/app/serve.go]: fixed

**Comment by sttts** [comment]
@cheftako @deads2k addressed your comments.

**Comment by sttts** [comment]
/assign @mikedanese 

For approval. Also compare https://github.com/kubernetes/kubernetes/pull/59814 to avoid this in the future for the owner-less generic controller-manager code (which should not stay in cmd/ for long, but that's another story).

**Pull request merged**